ADR/propagate-pull-secrets.md

```diff
@@ -0,0 +1,52 @@
+# 0. Support deploying images from Private Quay.io repositories
+
+Date: 2023-03-16
+
+## Status
+
+Provisional
+
+## Context
+
+* As a user, I would expect StoneSoup to generate a private Quay.io image repository if my source code repository is  private.
+* As a user, I would expect StoneSoup to gracefully deploy images that were pushed to private Quay.io repositories in every relevant environment.
+
+
+## Decision
+
+### Overview
+
+1. This proposal builds on the capabilties of the https://github.com/redhat-appstudio/image-controller which sets up a Quay.io image repository per `Component` and 'downloads' a robot account token into the user's namespace for builds to be able to push to it.
+
+2. This proposal builds on the capabilities of SPI's `RemoteSecret` API. See https://github.com/redhat-appstudio/book/pull/70/files
+
+
+#### When a new `Component` is created
+
+1. A new Quay.io image repository and robot account token is made available in the namespace that would only be able to "pull" the `Component`'s image.
+2. The image-controller labels the `Secret` with the name `Application`/`Component` the `Secret` is associated with. 
+
+
+#### When an expression of intent to deploy an `Application` to an `Environment` is made.
+
+
+1. Watch for creation of new `SnapshotEnvironmentBindings`.
+2. Query the 'pull' `Secret` generated by the image-controller. 
+3. Label the `Secret` with the name of the `Environment` ( ex, `spi.appstudio.redhat.com/target-environment: prod`) the `Secret` needs to be copied to, based on the `SnapshotEnvironmentBinding` CR.
+3. SPI generates a `RemoteSecret` CR referencing the 'pull' `Secret` above. ( See https://github.com/redhat-appstudio/book/pull/70/files )
+3. SPI reconciles the `RemoteSecret` CR, copies over the 'pull' `Secret` to the relevant cluster/namespace and links it to the `default` `serviceaccount`.
+
+TODO: Which Stonesoup controller should be doing (1), (2) and (3) 
```

Comment:
My two cents: this should be application-service, since it is already watching for and reacting to changes in the SnapshotEnvironmentBindings in order to write out updates to the gitops repo.

cc @elsony

Reply:
How often does the `Secret` changes? If we are only doing that immediately after the creation of `SnapshotEnvironmentBindings`, is there a reason why we don't include that information during the creation of the `SnapshotEnvironmentBindings` so we don't have extra logic to handle this?

Reply:
@sbose78, wdyt? I think the Secret will "normally" only be provisioned once when the Component is created - although it is possible we'll want to delete it and have it be reprovisioned manually - or, someday introduce automated rotation.

Reply:
@elsony Yeah, that's do-able as well. My proposal mostly intended to address a separation of concerns but doing it along with creation of `SnapshotEnvironmentBindings` works as well.

@ralphbean , Yeah, the Secret shouldn't change a lot. I'm hoping if/when the contents of the `Secret` does change, SPI should take care of getting it sync'd.